master/internal/api_trials.go

```diff
@@ -308,52 +310,95 @@ func (a *apiServer) GetTrialCheckpoints(
 	}
 
-	resp := &apiv1.GetTrialCheckpointsResponse{}
-	resp.Checkpoints = []*checkpointv1.Checkpoint{}
-
-	switch err := a.m.db.QueryProto("get_checkpoints_for_trial", &resp.Checkpoints, req.Id); {
-	case err == db.ErrNotFound:
-		return nil, status.Errorf(
-			codes.NotFound, "no checkpoints found for trial %d", req.Id)
-	case err != nil:
-		return nil,
-			errors.Wrapf(err, "error fetching checkpoints for trial %d from database", req.Id)
-	}
+	ext := db.QueryExt{}
 
-	a.filter(&resp.Checkpoints, func(i int) bool {
-		v := resp.Checkpoints[i]
+	// always filter by trial
+	ext = ext.Where("trial_id = ?", req.Id)
 
-		found := false
-		for _, state := range req.States {
-			if state == v.State {
-				found = true
-				break
+	if len(req.ValidationStates) > 0 {
+		var states []string
+		for _, vstate := range req.ValidationStates {
+			switch vstate {
+			case checkpointv1.State_STATE_ACTIVE:
+				states = append(states, "'ACTIVE'")
+			case checkpointv1.State_STATE_COMPLETED:
+				states = append(states, "'COMPLETED'")
+			case checkpointv1.State_STATE_ERROR:
+				states = append(states, "'ERROR'")
+			case checkpointv1.State_STATE_DELETED:
+				states = append(states, "'DELETED'")
+			default:
+				// XXX: throw an error to the user
 			}
 		}
+		// filter by state
+		ext = ext.Where("checkpoint_state IN (?)", bun.In(states))
+	}
+
+	// choose an ordering
+	asc := " ASC"
+	if req.OrderBy == apiv1.OrderBy_ORDER_BY_DESC {
+		asc = " DESC"
+	}
+	switch req.SortBy {
+	case apiv1.GetTrialCheckpointsRequest_SORT_BY_UNSPECIFIED:
+		// noop, use the default
+	case apiv1.GetTrialCheckpointsRequest_SORT_BY_BATCH_NUMBER:
+		// noop, this is the default
+	case apiv1.GetTrialCheckpointsRequest_SORT_BY_START_TIME:
+		// noop, this field doesn't exist anymore
+	case apiv1.GetTrialCheckpointsRequest_SORT_BY_END_TIME:
+		ext = ext.Order("report_time" + asc)
+	case apiv1.GetTrialCheckpointsRequest_SORT_BY_VALIDATION_STATE:
+		ext = ext.Order("validation_metrics != 'null'::jsonb" + asc)
+	case apiv1.GetTrialCheckpointsRequest_SORT_BY_STATE:
+		ext = ext.Order("state" + asc)
+	}
+
+	// secondary/default ordering is based on latest_batch
+	// XXX: the get_checkpoints_for_trial also had an ORDER BY clause for report time
+	ext = ext.Order("(metadata->>'latest_batch')::int8", asc)
+
+	resp := &apiv1.GetTrialCheckpointsResponse{}
+	// Count and paginate within a tx.
+	return resp, db.Bun.RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		total, err := ckpts.CountTx(ctx, tx, ext)
+		if err != nil {
+			return errors.Wrapf(err, "error counting checkpoints for trial %d from database", req.Id)
+		}
 
-		if len(req.States) != 0 && !found {
-			return false
+		// apply pagination
+		ext = ext.Limit(int(req.Limit))
+		ext = ext.Offset(int(req.Offset))
+
+		ckpts, err := ckpts.ListTx(ctx, tx, ext)
+		if err != nil {
+			return errors.Wrapf(err, "error fetching checkpoints for trial %d from database", req.Id)
+		}
```

Comment:
@ioga this is the code in question, where it's really nice to have this QueryExt object.

If we decide that separate CountTx and ListTx calls are not acceptable anyway, then we should just rm -rf the QueryExt.

Reply:
if we were to follow 1), i.e. simply inline it, then we'd be building a `bun.SelectQuery` from `db.NewSelect()` instead of `QueryExt`, call `query.Count(ctx)` to get the count, and then proceed with adding `Limit` and `Offset` to the same query.

Reply:
on a separate note, I think this transaction block doesn't do anything. You get "read committed" level by default, and these selects can get different committed data. 
you've got to set transaction to repeatable read if you want to get count & list results matching.

Reply:
> I think this transaction block doesn't do anything
Interesting.  So I'm fine with any of the following, in order I would prefer them:
- don't bother with transactions on pagination data, it seems racy anyway.  Pagination data will change from call to call so, eh.
- increase transaction level
- switch to one megaquery, as with get_experiments.  I see this as an optimization that complicates the code to require much more SQL skill (more than I have at least, and I'm having to write it).  IMO that's worth it only when we've identified this code as a bottleneck, otherwise I think it's a premature optimization.

Reply:
I'd vote for not bothering here.

Reply:
@stoksc So you like the closures and you don't like QueryExt (totally fine), but what about the megaquery vs two separate queries?

Reply:
two separate queries is fine.

Reply:
better, even.